lib/MySQL_HostGroups_Manager.cpp

```diff
@@ -7358,3 +7623,127 @@ MySrvC* MySQL_HostGroups_Manager::find_server_in_hg(unsigned int _hid, const std
+
+	for (auto& node : append) {
+		MySrvC* new_srv = insert_HGM(get_hostgroup_id(dest_type, node), node.srv);
+			
+		if (!new_srv) assert(0);
+			
+		node.srv = new_srv;
+		dest_nodes.push_back(node);
+	}
+}
+
+void MySQL_HostGroups_Manager::HostGroup_Server_Mapping::remove(Type type, size_t index) {
+
+	std::vector<Node>& nodes = mapping[type];
+
+	// ensure that we're not attempting to access out of the bounds of the container.
+	assert(index < nodes.size());
+
+	remove_HGM(nodes[index].srv);
+
+	//Swap the element with the back element, except in the case when we're the last element.
+	if (index + 1 != nodes.size())
+		std::swap(nodes[index], nodes.back());
+
+	//Pop the back of the container, deleting our old element.
+	nodes.pop_back();
+}
+
+void MySQL_HostGroups_Manager::HostGroup_Server_Mapping::clear(Type type) {
+
+	for (const auto& node : mapping[type]) {
+		remove_HGM(node.srv);
+	}
+
+	mapping[type].clear();
+}
+
+unsigned int MySQL_HostGroups_Manager::HostGroup_Server_Mapping::get_hostgroup_id(Type type, size_t index) const {
+
+	if (type == Type::WRITER)
+		return mapping[Type::WRITER][index].writer_hostgroup_id;
+	else if (type == Type::READER)
+		return mapping[Type::READER][index].reader_hostgroup_id;
+	else
+		assert(0);
+}
+
+unsigned int MySQL_HostGroups_Manager::HostGroup_Server_Mapping::get_hostgroup_id(Type type, const Node& node) const {
+
+	if (type == Type::WRITER)
+		return node.writer_hostgroup_id;
+	else if (type == Type::READER)
+		return node.reader_hostgroup_id;
+	else
+		assert(0);
+}
+
+MySrvC* MySQL_HostGroups_Manager::HostGroup_Server_Mapping::insert_HGM(unsigned int hostgroup_id, const MySrvC* srv) {
+
+	MyHGC* hostgroup_container = myHGM->MyHGC_lookup(hostgroup_id);
+
+	if (!hostgroup_container)
+		return NULL;
+
+	if (GloMTH->variables.hostgroup_manager_verbose) {
+		proxy_info("Creating new server in HG %d : %s:%d , gtid_port=%d, weight=%d, status=%d\n", hostgroup_id, srv->address, srv->port, srv->gtid_port, srv->weight, srv->status);
+	}
+
+	proxy_debug(PROXY_DEBUG_MYSQL_CONNPOOL, 5, "Adding new server %s:%d , weight=%d, status=%d, mem_ptr=%p into hostgroup=%d\n", srv->address, srv->port, srv->weight, srv->status, srv, hostgroup_id);
+
+	MySrvC* new_srv = new MySrvC(srv->address, srv->port, srv->gtid_port, srv->weight, srv->status, srv->compression,
+		srv->max_connections, srv->max_replication_lag, srv->use_ssl, srv->max_latency_us, srv->comment);
+
+	hostgroup_container->mysrvs->add(new_srv);
+
+	return new_srv;
+}
+
+void MySQL_HostGroups_Manager::HostGroup_Server_Mapping::remove_HGM(MySrvC* srv) {
+	proxy_warning("Removed server at address %p, hostgroup %d, address %s port %d. Setting status OFFLINE HARD and immediately dropping all free connections. Used connections will be dropped when trying to use them\n", (void*)srv, srv->myhgc->hid, srv->address, srv->port);
+	srv->status = MYSQL_SERVER_STATUS_OFFLINE_HARD;
+	srv->ConnectionsFree->drop_all_connections();
+}
```

Comment:
This is a picky one, I have seen this in previous file editions, probably it's an editor setting that doesn't add a new line at the end of file (typical of UNIX systems). If you could add it would be great.

Reply:
Done